### DISCUSSION TITLE
Content Security Policy 🛂

### DISCUSSION THREAD

**Discussion by StorytellerCZ** [meteor/meteor#11591] [Feature Requests]
For a while there has been a discussion about Content Security Policy and how to proceed with it.

https://github.com/meteor/meteor/issues/9689
https://github.com/meteor/meteor/issues/11424
https://github.com/meteor/meteor/issues/10704
And more in old feature requests.

I'm now concentrating the discussion here.

I'm proposing that we upgrade the existing packages to use [Helmet](https://github.com/helmetjs/helmet) under the hood and wire up all the additional options so that devs don't have to worry about those things like nonce generation and provide good defaults.

Thoughts?

**Comment by krstffr** [comment]
**Adding my use case/situation for your consideration**: I am currently using Helmet, and do not want to allow `unsafe-eval` (who would?), but would love to be able to use `dynamic-imports`, but currently can not due to this conflict.

**Reply by jankapunkt** [reply]
I wonder how others do that to overcome unsafe-eval? I only found polyfills or implementations with eval or function constructor, although I read some use a workaround with blob+nonce but I found no implementation on that.

**Reply by cormip** [reply]
Using nonce attributes does not permit scripts with unsafe evals to be run. See: https://csplite.com/csp/test152/

**Reply by italojs** [reply]
but cant we disable only `unsafe-eval` ? helmet is pretty flexible 
I understand this path solve a problem but create a new one, but IMO it's a smaller problem to solve

**Comment by cormip** [comment]
Pretty much every security scanner out there will flag `script-src: 'unsafe-eval'` as a significant security concern:

https://observatory.mozilla.org/
![image](https://user-images.githubusercontent.com/1108728/226116990-4bb014a8-aad7-44ac-a44b-c46c09a9da38.png)

OWASP Zap
![image](https://user-images.githubusercontent.com/1108728/226117341-285c730d-ff44-4ace-a43c-80f89c451204.png)

This needs to be addressed. It makes it difficult to pitch MeteorJS as a "very secure" platform, when it generates "mediocre" scores from the security scanners.

**Reply by paulincai** [reply]
https://forums.meteor.com/t/browserpolicy-content-disalloweval-still-gives-unsafe-eval/50336

It can be fixed by ... not using dynamic-imports ... Not a solution I can use though.

**Reply by jankapunkt** [reply]
How does webpack etc. solve this? They also Support dynamic Imports and I wonder if These Apps require unsafe eval for it as well?

**Comment by HemalR** [comment]
Is this by any chance something that is fixed in Meteor 3? As @cormip mentioned - the mediocre security scores generated by the inclusion of unsafe-eval can cause issues (and is doing so in our case). 

Are there any other workarounds?

**Reply by jankapunkt** [reply]
From what I have seen there is no fix for `dynamic-import` without `unsafe-eval` so far.

**Reply by StorytellerCZ** [reply]
I think that after Meteor 3.0 this is going to eventually become an increasing issue. I vote for CSP packages and this issue in particularly to be the security improvements focus for 3.1 or 3.2. I personally can help with modernizing the Meteor's CSP packages, but the `dynamic-import` without `unsafe-eval` issue is probably beyond me at this time.

**Reply by HemalR** [reply]
When you say 'CSP packages' are you referring to the `browser-policy` package? The csp I've implemented uses helmet and that seems to work reasonably well. Although I'm not clued in enough about this and maybe there is merit in having a Meteor specific package for this.

**Reply by StorytellerCZ** [reply]
Yes, the `browser-policy` packages. I would like to either integrate Helmet into those packages, or even better with some love figure out an approach to solve the above mentioned issue and do other optimizations specific for Meteor. Honestly, right now I'm handling CSP through the WebApp package easily, so these packages need to add significant value to justify their use (or at the very least make it so that people don't have to think about CSP and get the security that comes when having it properly configured).

**Comment by italojs** [comment]
The discussion was moved to the Meteor forum [here](https://forums.meteor.com/t/content-security-policy-csp-implementation-dynamic-imports-challenge/64443).
